drivers/serial/uart_ns16550.c

```diff
@@ -1721,16 +1760,34 @@ static int uart_ns16550_rx_enable(const struct device *dev, uint8_t *buf, const
 		goto out;
 	}
 
+	rx_dma_params->async_enabled = true;
 	rx_dma_params->timeout_us = timeout_us;
 	rx_dma_params->buf = buf;
 	rx_dma_params->buf_len = len;
 
 #if defined(CONFIG_UART_NS16550_INTEL_LPSS_DMA)
 	ns16550_outword(config, MST(dev), UNMASK_LPSS_INT(rx_dma_params->dma_channel));
 #else
-	ns16550_outbyte(config, IER(dev),
-			(ns16550_inbyte(config, IER(dev)) | IER_RXRDY));
-	ns16550_outbyte(config, FCR(dev), FCR_FIFO);
+	if ((timeout_us != SYS_FOREVER_US) && (timeout_us != 0)) {
+		/* IRQ is needed to make character timeout working */
+		ns16550_outbyte(config, IER(dev),
+				(ns16550_inbyte(config, IER(dev)) | IER_RXRDY));
+		irq_enable(data->irq);
+	}
+	uint32_t fcr_reg = FCR_FIFO;
+#if CONFIG_DMA_PL330
+	/*
+	 * Program FIFO: enabled
+	 *   RX generates the interrupt when 1 byte in FIFO
+	 *   TX generates the interrupt when FIFO is 1/4 full
+	 *   Clear TX and RX FIFO if system is not resuming
+	 */
+	fcr_reg += FCR_TX_FIFO_QUARTER | FCR_FIFO_1 | FCR_RCVRCLR | FCR_XMITCLR;
```

Comment:
Do you actually mean this?
```suggestion
	fcr_reg |= FCR_TX_FIFO_QUARTER | FCR_FIFO_1 | FCR_RCVRCLR | FCR_XMITCLR;
```

Reply:
no, `+=` is more robust way to handle that because values could be different in size.

Reply:
Integer promotion happens for bitwise OR so I'm not convinced about the added robustness.
One could also argue that using addition can be less robust... if the fcr_reg had one of the flags (which we are now setting) already set earlier in the code then bitwise OR would give correct result while addition fails.

Reply:
yeah, in that sense I agree. I can modify this even it does not really matter which way this is done. I recall some cases where GCC messed up flag calculation of different types but that could be enhanced nowadays :)
Will be changed...

Reply:
done